### ISSUE TITLE
[azure-iot-sdk] DeviceClient OpenAsync retries forever if connectionstring contains "invalid" host

### ISSUE THREAD

**Issue by goetae** [Azure/azure-iot-sdk-csharp#2181]
# Context

- **OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64,
- **Application's .NET Target Framework :** .net472, netcore3.1, netcore 5.0
- **Device:** Laptop
- **SDK version used:** 1.38

## Description of the issue
Providing a connectionstring with an "invalid" DNS e.g. not exiting DNS (any random string) or DNS without IOT Hub endpointb endpoint e.g. 
  - myiothubinstance.azure-devices.**com** instead of myiothubinstance.azure-devices.**net**
  - providing any other DNS e.g. "microsoft.com" or non existing DNS like "thisdomainistnotregisteredatall.com"
 
e.g. "HostName=microsoft.com;DeviceId=mydeviceid;SharedAccessKey=mykey"
the client will retry "forever" - at least for more than 30 minutes 

## Code sample exhibiting the issue

My code is based on various samples, in this case on the [reconnect sample](https://github.com/Azure-Samples/azure-iot-samples-csharp/blob/9561e18bd5c0edc703062f80b29128756f14a77a/iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs#L160)
I tested the behavior with the sample code as well

## Console log of the issue
```
2021-10-01T10:00:25.0598883>> Information - Supplied with 2 connection string(s).
2021-10-01T10:00:25.4096189>> Information - Using Mqtt transport.
2021-10-01T10:00:29.3347818>> Information - Sample execution started, press Control+C to quit the sample.
2021-10-01T10:00:39.7263927>> Debug - Attempting to initialize the client instance, current status=Disconnected
2021-10-01T10:00:40.0220956>> Debug - Initialized the client instance.
2021-10-01T10:00:40.0252718>> Information - Operation name = OpenConnection, attempt = 0, status = executing.
2021-10-01T10:00:40.6483759>> Debug - Connection status changed: status=Disconnected_Retrying, reason=Communication_Error
2021-10-01T10:00:40.6496636>> Debug - ### The DeviceClient is retrying based on the retry policy. Do NOT close or open the DeviceClient instance
2021-10-01T10:27:24.8993145>> Information - Sample execution cancellation requested; will exit.
2021-10-01T10:27:31.9926532>> Debug - Connection status changed: status=Disconnected, reason=Retry_Expired
2021-10-01T10:27:31.9936327>> Warning - ### The DeviceClient has been disconnected because the retry policy expired.
```
waiting more than 20 minutes without any changes, then cancelled / closed the application

**Comment by azabbasi** [comment]
I am a bit confused about the issue description, are you trying to provide a device connection string using a DNS like microsoft.com? you should be using the host name of your IoT hub.
`HostName=[Your IoT Hub Host Name];DeviceId=]YourDeviceId];SharedAccessKey=[yourAccessKey]`

**Comment by goetae** [comment]
hello,

I just wanted to point out that any **typo** in DNS will result in an "endless retry". 
I just tried to make an extreme example by testing the connection with a total different DNS.

I create a MSI,WIX installer with some IOT/DPS presets, but a user might change the endpoints and if there is a typo and the client connects to an "invalid" DNS, the code cannot react as the DeviceClient will retry forever?! or at least for more than 20 minutes. I wonder when I get any notification on an unsuccessful connect.

I personally would like to have the client to fail fast on any connectivity issue. I'm currently migrating a simple MQTT client application to Azure IOT and I do some of my common tests e.g. disabling FW ports, change DNS on network level, disable network and as well configuration issues / typos. 

I do the same tests with other azure clients e.g. cosmosdb, eventhub, if there is anything wrong with DNS they fail on initialization e.g. "System.Net.Http.HttpRequestException: No such host is known." 

I also experienced issues with the ConnectionStatusChangeHandler not triggering always, also leaving the client in a "unusuable" state, as mentioned by others, I might open anyother issue on this but I cannot provide reproduction steps as for now :(
